Mathlib/CategoryTheory/GradedObject/Unitor.lean

```diff
@@ -252,6 +252,105 @@ lemma mapBifunctorRightUnitor_naturality :
+@[reducible]
+def ρ₂₃ : BifunctorComp₂₃IndexData r where
+  I₂₃ := I₃
+  p := τ.p₂₃
+  q := π
+  hpq := τ.hp₂₃
+
+end TriangleIndexData
+
+end
+
+section Triangle
+
+variable {C₁ C₂ C₃ D I₁ I₂ I₃ J : Type*} [Category C₁] [Category C₂] [Category C₃] [Category D]
+  [Zero I₂] [DecidableEq I₂] [HasInitial C₂]
+  {F₁ : C₁ ⥤ C₂ ⥤ C₁} {F₂ : C₂ ⥤ C₃ ⥤ C₃} {G : C₁ ⥤ C₃ ⥤ D}
+  (associator : bifunctorComp₁₂ F₁ G ≅ bifunctorComp₂₃ G F₂)
+  (X₂ : C₂) (e₁ : F₁.flip.obj X₂ ≅ 𝟭 C₁) (e₂ : F₂.obj X₂ ≅ 𝟭 C₃)
+  [∀ (X₁ : C₁), PreservesColimit (Functor.empty.{0} C₂) (F₁.obj X₁)]
+  [∀ (X₃ : C₃), PreservesColimit (Functor.empty.{0} C₂) (F₂.flip.obj X₃)]
+  {r : I₁ × I₂ × I₃ → J} {π : I₁ × I₃ → J}
+  (τ : TriangleIndexData r π)
+  (X₁ : GradedObject I₁ C₁) (X₃ : GradedObject I₃ C₃)
+  [HasMap (((mapBifunctor F₁ I₁ I₂).obj X₁).obj ((single₀ I₂).obj X₂)) τ.p₁₂]
+  [HasMap (((mapBifunctor G I₁ I₃).obj
+    (mapBifunctorMapObj F₁ τ.p₁₂ X₁ ((single₀ I₂).obj X₂))).obj X₃) π]
+  [HasMap (((mapBifunctor F₂ I₂ I₃).obj ((single₀ I₂).obj X₂)).obj X₃) τ.p₂₃]
+  [HasMap (((mapBifunctor G I₁ I₃).obj X₁).obj
+      (mapBifunctorMapObj F₂ τ.p₂₃ ((single₀ I₂).obj X₂) X₃)) π]
+  [HasGoodTrifunctor₁₂Obj F₁ G τ.ρ₁₂ X₁ ((single₀ I₂).obj X₂) X₃]
+  [HasGoodTrifunctor₂₃Obj G F₂ τ.ρ₂₃ X₁ ((single₀ I₂).obj X₂) X₃]
+  [HasMap (((mapBifunctor G I₁ I₃).obj X₁).obj X₃) π]
+  (triangle : ∀ (X₁ : C₁) (X₃ : C₃), ((associator.hom.app X₁).app X₂).app X₃ ≫
+    (G.obj X₁).map (e₂.hom.app X₃) = (G.map (e₁.hom.app X₁)).app X₃)
+
+lemma mapBifunctor_triangle :
+    (mapBifunctorAssociator associator τ.ρ₁₂ τ.ρ₂₃ X₁ ((single₀ I₂).obj X₂) X₃).hom ≫
+    mapBifunctorMapMap G π (𝟙 X₁) (mapBifunctorLeftUnitor F₂ X₂ e₂ τ.p₂₃ τ.h₃ X₃).hom =
+      mapBifunctorMapMap G π (mapBifunctorRightUnitor F₁ X₂ e₁ τ.p₁₂ τ.h₁ X₁).hom (𝟙 X₃) := by
+  rw [← cancel_epi ((mapBifunctorMapMap G π
+    (mapBifunctorRightUnitor F₁ X₂ e₁ τ.p₁₂ τ.h₁ X₁).inv (𝟙 X₃)))]
+  ext j i₁ i₃ hj
+  simp only [categoryOfGradedObjects_comp, ι_mapBifunctorMapMap_assoc,
+    mapBifunctorRightUnitor_inv_apply, Functor.id_obj, Functor.flip_obj_obj, Functor.map_comp,
+    NatTrans.comp_app, categoryOfGradedObjects_id, Functor.map_id, id_comp, assoc,
+    ι_mapBifunctorMapMap]
+  congr 2
+  rw [← ιMapBifunctor₁₂BifunctorMapObj_eq_assoc F₁ G τ.ρ₁₂ _ _ _ i₁ 0 i₃ j
+    (by rw [τ.r_zero, hj]) i₁ (by simp), ι_mapBifunctorAssociator_hom_assoc,
+    ιMapBifunctorBifunctor₂₃MapObj_eq_assoc G F₂ τ.ρ₂₃ _ _ _ i₁ 0 i₃ j
+    (by rw [τ.r_zero, hj]) i₃ (by simp), ι_mapBifunctorMapMap]
+  dsimp
+  rw [Functor.map_id, NatTrans.id_app, id_comp,
+    ← Functor.map_comp_assoc, ← NatTrans.comp_app_assoc, ← Functor.map_comp,
+    ι_mapBifunctorLeftUnitor_hom_apply F₂ X₂ e₂ τ.p₂₃ τ.h₃ X₃ i₃,
+    ι_mapBifunctorRightUnitor_hom_apply F₁ X₂ e₁ τ.p₁₂ τ.h₁ X₁ i₁]
+  dsimp
+  simp only [Functor.map_comp, assoc, NatTrans.comp_app, ← triangle (X₁ i₁) (X₃ i₃)]
+  erw [← NatTrans.naturality_app_assoc]
+  rfl
```

Comment:
Would be great if we could get rid of this defeq abuse. We can still merge this, I just feel that every case where something like this is necessary makes the API a bit harder to learn, because you can no longer see all the required tools by studying the available lemmas, you also need to keep all of the definitions in your head.

Reply:
If you are referring to the `rfl`, `dsimp` was also closing the goal. If you are referring to the `erw`, I did not manage to make a significant improvement. The end of the proof is now:
```lean
  simp only [← assoc]
  congr 2
  symm
  apply NatTrans.naturality_app (associator.hom.app (X₁ i₁))
```